### PR TITLE
Wire Webauthn to login endpoints

### PR DESCRIPTION
**Pull request by codingllama**
Allow login using Webauthn instead of U2F.

Installs with second_factor:on will automatically support Webauthn, generating
challenges for any U2F devices and allowing Webauthn to be used throughout the
entire login process.

Web REST API modified as follows:

* Added /webapi/mfa/login/begin (1st login step)
* Added /webapi/mfa/login/finish (2nd login step, SSH)
* Added /webapi/mfa/login/finishsession (2nd login step, web session)
* U2F endpoints deprecated, but support full breadth of MFA login options
  (/webapi/u2f/signrequest, /webapi/u2f/certs and /webapi/u2f/sessions,
  respectively)

Auth REST API modified as follows:

* Added /:version/mfa/users/:user/login/begin (1st login step)
* ssh/authenticate and web/authenticate now support Webauthn (2nd login step)
* U2F endpoints deprecated, but support full breadth of MFA options as well
  (/:version/u2f/users/:user/sign)

RFD: https://github.com/gravitational/teleport/pull/7808

### CODE REVIEW

File: lib/auth/auth.go [hunk 1]
```diff
@@ -2446,56 +2454,57 @@ func (a *Server) mfaAuthChallenge(ctx context.Context, user string, u2fStorage u
...
+		enableU2F = true
+	case constants.SecondFactorWebauthn:
+		enableWebauthn = true
+	case constants.SecondFactorOn, constants.SecondFactorOptional:
+		enableTOTP, enableU2F, enableWebauthn = true, true, true
 	default:
-		// Other AuthPreference types don't restrict us to a single MFA type,
-		// send all challenges.
-		enableTOTP, enableU2F = true, true
+		return nil, trace.Errorf("unexpected second_factor value: %s", apref.GetSecondFactor())
```
Comment (zmb3) [inline lib/auth/auth.go:2468]: Would this be an appropriate place to use `trace.BadParameter`?

File: lib/auth/auth.go [hunk 2]
```diff
@@ -2511,9 +2520,49 @@ func (a *Server) mfaAuthChallenge(ctx context.Context, user string, u2fStorage u
...
+	U2F      []*types.MFADevice
+	Webauthn []*types.MFADevice
+}
+
+func groupByDeviceType(devs []*types.MFADevice, groupU2F, groupWebauthn bool) devicesByType {
+	res := devicesByType{}
+	for _, dev := range devs {
+		switch dev.Device.(type) {
+		case *types.MFADevice_Totp:
+			res.TOTP = true
```
Comment (zmb3) [inline lib/auth/auth.go:2550]: Why do we only have a single bool flag for TOTP devices?

File: lib/auth/methods.go [hunk 3]
```diff
@@ -125,84 +127,112 @@ func (s *Server) authenticateUser(ctx context.Context, req AuthenticateUserReque
...
+		// MFA.
+		log.Warningf("MFA bypass attempt by user %q, access denied.", req.Username)
+		return nil, trace.AccessDenied("missing second factor")
+	}
+	if err = s.WithUserLock(req.Username, func() error {
+		return s.checkPasswordWOToken(req.Username, req.Pass.Password)
+	}); err != nil {
+		// provide obscure message on purpose, while logging the real
+		// error server side
+		log.Debugf("Failed to authenticate: %v.", err)
```
Comment (zmb3) [inline lib/auth/methods.go:198]: Include the username in the debug log?

File: lib/auth/webauthn/messages.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	// leaking the duo-labs/webauthn dependency.
+	// The nesting of types is identical to protocol.CredentialAssertionResponse.
+
+	PublicKeyCredential
+	AssertionResponse AuthenticatorAssertionResponse `json:"response"`
+}
+
+type PublicKeyCredential struct {
+	Credential
+	RawID      protocol.URLEncodedBase64              `json:"rawId"`
```
Comment (zmb3) [inline lib/auth/webauthn/messages.go:38]: Does "raw" mean still base64 encoded?

**Comment by codingllama** [comment]
Taking reviewers off temporarily, I'm adding a few more tests.

File: lib/auth/auth.go (hunk 1 above)
Comment (codingllama) [reply lib/auth/auth.go:2468]: Yes, we apparently default to `trace.BadParameter`. I'll fix it.

File: lib/auth/auth.go (hunk 2 above)
Comment (codingllama) [reply lib/auth/auth.go:2550]: This is the legacy behavior - TOTP is either enabled or not.

File: lib/auth/webauthn/messages.go (hunk 4 above)
Comment (codingllama) [reply lib/auth/webauthn/messages.go:38]: Just double-checked, it is correct. The standard defines [rawId as an ArrayBuffer](https://www.w3.org/TR/webauthn-2/#iface-pkcredential), but I suppose one must encode it anyway for transport.

File: lib/auth/methods.go (hunk 3 above)
Comment (codingllama) [reply lib/auth/methods.go:198]: Done.

**Comment by codingllama** [comment]
PTAL.

Apologies for the back-and-forth, I decided to add a bit more tests in the end. CI issues should also be fixed now (licenses and grpcserver_test).

**Comment by codingllama** [comment]
PTAL?

Added even more testing (SSH and Web login with various MFA devices - https://github.com/gravitational/teleport/pull/8094/commits/4853995fd96eeaafcfa0f9adbe5d5ea873e07610). I think I'm happy now, so I'll stop poking here. Thanks. :)

**Review by quinqu — APPROVED** [review]

**Review by zmb3 — APPROVED** [review]

**Comment by codingllama** [comment]
Thanks everyone.

**Review by klizhentas — APPROVED** [review]

File: lib/web/apiserver.go [hunk 5]
```diff
@@ -338,13 +338,17 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*RewritingHandler, error) {
...
-	h.POST("/webapi/u2f/sessions", httplib.MakeHandler(h.createSessionWithU2FSignResponse))
-	h.POST("/webapi/u2f/certs", httplib.MakeHandler(h.createSSHCertWithMFAChallengeResponse))
+	h.POST("/webapi/u2f/signrequest", httplib.MakeHandler(h.mfaLoginBegin))
+	h.POST("/webapi/u2f/sessions", httplib.MakeHandler(h.mfaLoginFinishSession))
+	h.POST("/webapi/u2f/certs", httplib.MakeHandler(h.mfaLoginFinish))
 
 	// MFA related endpoints
+	h.POST("/webapi/mfa/login/begin", httplib.MakeHandler(h.mfaLoginBegin))
+	h.POST("/webapi/mfa/login/finish", httplib.MakeHandler(h.mfaLoginFinish))
+	h.POST("/webapi/mfa/login/finishsession", httplib.MakeHandler(h.mfaLoginFinishSession))
```
Comment (r0mant) [inline lib/web/apiserver.go:351]: Sorry just looking at this, but IIRC our HTTP server is deprecated. All new handles should be a part of GRPC API. Do these have to be HTTP?

Comment (codingllama) [reply lib/web/apiserver.go:351]: These are meant for login, that's why we are keeping the HTTP endpoints. I describe them in the [WebAuthn RFD](https://github.com/gravitational/teleport/pull/7808); I believe that, by and large, reviewers are OK with the proposal.

I've also raised https://github.com/gravitational/teleport/issues/7966 to eventually move login to gRPC. I intent to follow up, but I didn't want to change too much at the same time for various reasons (including the fact that I'm still getting to know the codebase better).

**Review by r0mant — APPROVED** [review]

File: lib/web/apiserver.go (hunk 5 above)
Comment (r0mant) [reply lib/web/apiserver.go:351]: Oh this is lib/web, this is fine then.

**Pull request merged**